oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java

```diff
@@ -118,6 +99,84 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
      */
     @Override
     public final CompletionStage<GrantedOAuth2AccessToken> getAccessToken() {
-        return tokenManager.get();
+        final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
+        final CompletableFuture<GrantedOAuth2AccessToken> tokenFuture = this.tokenFuture;
+
+        tokenFuture.handle((token, ignored) -> {
+            if (isValidToken(token)) {
+                future.complete(token);
+                return null;
+            }
+
+            final Supplier<CompletionStage<GrantedOAuth2AccessToken>> tokenIssuingFunc;
+            if (tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
+                tokenIssuingFunc = token != null && token.isRefreshable() ? () -> refreshAccessToken(token)
+                                                                          : () -> issueAccessToken(token);
+            } else {
+                tokenIssuingFunc = () -> this.tokenFuture;
+            }
+
+            final CompletionStage<GrantedOAuth2AccessToken> tokenIssuingFuture = RequestContext.mapCurrent(
+                            ctx -> ctx.makeContextAware(tokenIssuingFunc.get()), tokenIssuingFunc);
+            tokenIssuingFuture.handle((newToken, cause) -> {
+                if (cause != null) {
+                    future.completeExceptionally(cause);
+                } else {
+                    future.complete(newToken);
+                    if (saveTokenFunc != null) {
+                        // suppress 'ReturnValueIgnored' warning
+                        final CompletableFuture<Void> unused = saveTokenFunc.apply(newToken);
```

Comment:
I wanted to force async execution by interface so that the API become less error-prone.
I am down if you think `Consumer` interface and documentation will do the work.